lib/Echidna/Output/Source.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE Rank2Types #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TupleSections #-}
+
+module Echidna.Output.Source where
+
+import Control.Lens
+import Data.Maybe (fromMaybe, mapMaybe)
+import Data.Text (Text, unlines, pack, unpack)
+import Data.Text.Encoding (decodeUtf8)
+import Data.Text.IO (writeFile)
+import Data.List (nub)
+
+import EVM.Solidity (stripBytecodeMetadata, SourceCache, SrcMap, SolcContract, sourceLines, sourceFiles, runtimeCode, runtimeSrcmap, creationSrcmap)
+import EVM.Debug (srcMapCodePos)
+import Prelude hiding (unlines, writeFile)
+
+import qualified Data.Vector as V
+
+import qualified Data.Map as M
+import qualified Data.Set as S
+
+import Echidna.Exec 
+import Echidna.Types.Tx
+
+
+type FilePathText = Text
+
+saveCoveredCode :: Maybe FilePath -> SourceCache -> [SolcContract] -> CoverageMap -> IO ()
+saveCoveredCode (Just d) sc cs s = let fn = d ++ "/covered.txt"
+                                       cc = ppCoveredCode sc cs s 
+                                   in writeFile fn cc
+saveCoveredCode Nothing  _  _  _ = pure ()
+
+
+-- | Pretty-print the covered code
+ppCoveredCode :: SourceCache -> [SolcContract] -> CoverageMap -> Text
+ppCoveredCode sc cs s | s == mempty = "Coverage map is empty"
+                      | otherwise   =
+  let allLines = M.toList $ view sourceLines sc
+  -- ^ Collect all the possible lines from all the files
+      findFile k = fst $ M.findWithDefault ("<no source code>", mempty) k (view sourceFiles sc)
```

Comment:
`(view sourceFiles sc)` -> `(sc ^. sourceFiles)`

also, is it possible that there is no source code for a file?

Reply:
> (view sourceFiles sc) -> (sc ^. sourceFiles)

Fixed.

> also, is it possible that there is no source code for a file?

That should be a bug, but it depends from the compilation process, so could be.